### ISSUE TITLE
Option to configure TCP Keepalive

### ISSUE THREAD

**Issue by specterz** [timberio/vector#4816]
### Current Vector Version
0.10

### Use-cases
Hi Team,

It would be beneficial to have a option to configure TCP Keepalives (optional).

This would be benefical in environments, where firewalls will close TCP connections due to Idle timeout.

I experienced this, when using Syslog-TLS input.

### Attempted Solutions

The present solution would be to increase firewall idle timeout. However, given 1 hr is timeout standard in most enterprise firewall, TCP keepalive makes the more sense.

### Proposal
Option to configure TCP Keepalive per input in vector.toml

### References
https://discord.com/channels/742820443487993987/746070591097798688/771775364170383411

**Comment by binarylogic** [comment]
cc @jamtur01 for up coming work.

**Comment by pablosichert** [comment]
Looks like the TCP keepalive information needs to be propagated to [`set_keepalive`](https://docs.rs/tokio/0.2.13/tokio/net/struct.TcpStream.html#method.set_keepalive) in Tokio's [`TcpStream`](https://docs.rs/tokio/0.2.13/tokio/net/struct.TcpStream.html).

Unfortunately, the option is missing beginning from `0.3.x` and will be reimplemented in future versions. Upstream issue: https://github.com/tokio-rs/tokio/issues/3109.

The new [`TcpKeepalive`](https://github.com/tokio-rs/mio/pull/1385/files#diff-114840f4b3c93a2a8f7bbc6229a87d8e64879ba5fee9d69e1fb717b0f8ed04f6R26-R46) configuration will also include `interval` and `retries` options, while on `0.2x.` (which is currently used) we can only set the keepalive `time`.

I'd suggest we implement the keepalive `time` option with the current tokio runtime, as it covers most use cases. For the other options another issue with blockers on the upstream feature and runtime upgrade would be suitable.

**Comment by jamtur01** [comment]
@lukesteensen @Hoverbear Thoughts ^^^

**Comment by specterz** [comment]
Hey guys, basic TCP keepalive functionality will help us big time.

**Comment by binarylogic** [comment]
> I'd suggest we implement the keepalive time option with the current tokio runtime, as it covers most use cases. For the other options another issue with blockers on the upstream feature and runtime upgrade would be suitable.

@pablosichert that sounds reasonable. Could you open the other issue and reference the blocking issues? We might decide to contribute upstream to unblock work.

**Comment by Hoverbear** [comment]
@pablosichert Well researched! =D I agree.

**Comment by jamtur01** [comment]
@pablosichert Ditto - let's implement based on `time` and go from there.
